sql: make Fatal invalid state transition ReportOrPanic

asubiotto opened the pull request:
It was unnecessary to panic in this case. The log message in MoveToDraining has
also been augmented to show the stack trace of the error that MoveToDraining
was called with.

This commit also reworks the Materializer behavior that led to this invalid
state transition.

Release note: None

Closes #53880

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54442)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54442#-:-MHMkvYI30eYStwQ6zKw:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/materializer.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54442#-MHMhApL8WEt9PA9Bt-k:-MHMhApL8WEt9PA9Bt-l:b-t16tvx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdada4dfa0269e64627f9a6ac665866108425ec3/pkg/sql/colexec/materializer.go#L52)):*
> ```Go
> 	row rowenc.EncDatumRow
> 
> 	// ouputRows stores the returned results of next() to be passed through an
> ```

nit: `s/ouputRows/outputRow/`.
___
*[pkg/sql/colexec/materializer_test.go, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54442#-MHMkBK-2PCPdTHc5URB:-MHMkBK0El34T57ZzRDV:bfgvgwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdada4dfa0269e64627f9a6ac665866108425ec3/pkg/sql/colexec/materializer_test.go#L206)):*
> ```Go
> 		0,   /* processorID */
> 		nil, /* input */
> 		nil,
> ```

super nit: a missing inlined comment.
___
*[pkg/sql/colexec/materializer_test.go, line 216 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54442#-MHMjivDDJ8c380NLcHl:-MHMjivDDJ8c380NLcHm:b-9pkwoi) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdada4dfa0269e64627f9a6ac665866108425ec3/pkg/sql/colexec/materializer_test.go#L216)):*
> ```Go
> 
> 	// Call ConsumerDone.
> 	m.ConsumerDone()
> ```

nit: should we call `m.Start`? I think the contract requires it.
___
*[pkg/sql/colexec/materializer_test.go, line 219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54442#-MHMkP2m4EzQvXn7pEbv:-MHMkP2m4EzQvXn7pEbw:b-gdgrs3) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdada4dfa0269e64627f9a6ac665866108425ec3/pkg/sql/colexec/materializer_test.go#L219)):*
> ```Go
> 	// We expect Next to panic since DrainMeta panics are currently not caught by
> 	// the materializer and it's not clear whether they should be since
> 	// implementers of DrainMeta do not return errors as panics.
> ```

That's an interesting point. On the one hand, it's true that we don't use the panic-catch mechanism of error propagation in `DrainMeta`, so we should not be catching the panics, but on the other hand, the stability would improve slightly if we did catch them. I think we should keep the current status quo and maybe re-evaluate this if we see another instance of such panic.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
TFTR

bors r=yuzefovich

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54442#-:-MHQJX1g7C9YO35ifZCe:bsvne40)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/materializer_test.go, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54442#-MHMkBK-2PCPdTHc5URB:-MHQHMBP9iJvWWHHL9h4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdada4dfa0269e64627f9a6ac665866108425ec3/pkg/sql/colexec/materializer_test.go#L206)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: a missing inlined comment.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/materializer_test.go, line 219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54442#-MHMkP2m4EzQvXn7pEbv:-MHQHODL3xJ48rNHbVZO:bf8xl5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdada4dfa0269e64627f9a6ac665866108425ec3/pkg/sql/colexec/materializer_test.go#L219)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

That's an interesting point. On the one hand, it's true that we don't use the panic-catch mechanism of error propagation in `DrainMeta`, so we should not be catching the panics, but on the other hand, the stability would improve slightly if we did catch them. I think we should keep the current status quo and maybe re-evaluate this if we see another instance of such panic.
</blockquote></details>

Stability is paramount, so I initially added a catch. The thing is that if it's due to a nil pointer for example, we get into this infinite loop where we return the nil pointer as an error then attempt to drain again. The other approach is to simply say "DrainMeta panicked, this is pretty bad, so let's just return the panic as an error and not drain anymore". The problem is that the outbox also needs to be updated since it's not only the materializer that will be the "root" drainer. This starts to look like too big a change so I think it's ok to go down this route because we're making the current situation better by showing the nil pointer stack trace (even though it does crash the server, but this issue crashed the server before anyway) so that we can more easily fix that next time.


<!-- Sent from Reviewable.io -->

Commit:
sql: make Fatal invalid state transition ReportOrPanic

It was unnecessary to panic in this case. The log message in MoveToDraining has
also been augmented to show the stack trace of the error that MoveToDraining
was called with.

This commit also reworks the Materializer behavior that led to this invalid
state transition.

Release note: None

asubiotto commented:
bors r=yuzefovich

yuzefovich reviewed (commented):
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54442)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/colexec/materializer.go, line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54442#-MHReRCm37j3OaXCfEvR:-MHReRCm37j3OaXCfEvS:b-eimrb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/86130f26679a6a0829b9807ea67dab59dd34f133/pkg/sql/colexec/materializer.go#L52)):*
> ```Go
> 	row rowenc.EncDatumRow
> 
> 	// ouputRow stores the returned results of next() to be passed through an
> ```

nit: `s/ouputRow/outputRow/`.
___
*[pkg/sql/colexec/materializer_test.go, line 219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54442#-MHMkP2m4EzQvXn7pEbv:-MHReb309Oj28xDrfQRc:b-l0z3xc) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdada4dfa0269e64627f9a6ac665866108425ec3/pkg/sql/colexec/materializer_test.go#L219)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Stability is paramount, so I initially added a catch. The thing is that if it's due to a nil pointer for example, we get into this infinite loop where we return the nil pointer as an error then attempt to drain again. The other approach is to simply say "DrainMeta panicked, this is pretty bad, so let's just return the panic as an error and not drain anymore". The problem is that the outbox also needs to be updated since it's not only the materializer that will be the "root" drainer. This starts to look like too big a change so I think it's ok to go down this route because we're making the current situation better by showing the nil pointer stack trace (even though it does crash the server, but this issue crashed the server before anyway) so that we can more easily fix that next time.
</blockquote></details>

I see, thanks for explaining.


<!-- Sent from Reviewable.io -->
